proposals/interface-types/Explainer.md

```diff
@@ -435,69 +412,152 @@ using the previously-introduced `string-to-memory` lowering instruction.
+instructions like [`memory.copy`] to copy directly from memory to memory.
+This partial evaluation will be formalized as a set of provably-equivalent
+rewrite rules which engines can use to *predictably* optimize adapter
+function code into equivalent core wasm code (at instantiation time, when
+client and provider are known).
+
+For example, the above two adapted modules can be merged and rewritten into the
+following core wasm module (using the [multi-memory] proposal and the `let`
+blocks introduced by the [function references] proposal):
+
+```wasm
+(module
+  (memory $mem1 1)
+  (memory $mem2 1)
+  (func $mem1Malloc (param i32) (result i32) ...)
+  (func $mem1Free (param i32) ...)
+  (func $mem2Malloc (param i32) (result i32) ...)
+  ...
+  (func $validateUtf8 (param i32 i32) ... defined by Unicode spec ...)
+  ...
+  (func $get_ (param i32 i32) (result i32 i32) ...)
+  (func $get (param $keySrc i32) (param $keyLen i32) (result i32 i32)
+    (call $validateUtf8 (local.get $keySrc) (local.get $keyLen))
+    (call $mem1Malloc (local.get $keyLen))
+    let (local $keyDst i32) (result i32 i32)
+      (memory.copy "mem2" "mem1" (local.get $keyDst) (local.get $keySrc) (local.get $keyLen))
+
+      (call $get_ (local.get $keyDst) (local.get $keyLen))
+
+      let (local $valSrc i32) (local $valLen i32) (result i32 i32)
+        (call $validateUtf8 (local.get $keySrc) (local.get $keyLen))
+        (call $mem2Malloc (local.get $valLen))
+        let (local $valDst i32) (result i32 i32)
+          (memory.copy "mem1" "mem2" (local.get $valDst) (local.get $valSrc) (local.get $valLen))
+
+          (call $mem1Free (local.get $valSrc))
+
+          local.get $valDst
+          local.get $valLen
+        end
+      end
+    end
+  )
+  (func $randomCode
+    ...
+    call $get
+    ...
+  )
+)
+```
+
+Some interesting things to notice in this code are:
+* For the special case of strings, in addition to copying memory, UTF-8
+  validation is required at the interface boundary since the engine has no way
+  to know that the source bytes are already valid.
+* While the original client and provider modules needed to export memory and
+  various functions so that they can be used by the containing adapted module,
+  the rewritten module can keep these encapsulated and use module-internal
+  references.
+* The `defer-call-export` in the original adapted module has been
+  rewritten into a plain `call` at the end of the rewritten function scope.
+  `defer-call-export` can be thought of as a lazily-evaluated call, with
+  evaluation occurring at the end of its (post-rewrite) enclosing scope. Thus,
+  `defer-call-export` is naturally aligned with lazy lifting, ensuring that
+  memory is still allocated when it is read.
+
 
 ### Strings with the GC proposal
 
-In preceding [shared-nothing example](#shared-nothing-linking-example),
+In the preceding [shared-nothing example](#shared-nothing-linking-example),
```

Comment:
Should the `#tag` here be updated to `#lifting-and-lowering`?

Reply:
Thanks!  Fixed.